src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/Object.WriteTests.cs

```diff
@@ -146,5 +148,71 @@ public static void WriteObjectWithNumberHandling()
             var options = new JsonSerializerOptions { NumberHandling = JsonNumberHandling.AllowReadingFromString };
             JsonSerializer.Serialize(new object(), options);
         }
+
+        /// <summary>
+        /// This test is constrained to run on Windows and MacOSX because it causes
+        /// problems on Linux due to the way deferred memory allocation works. On Linux, the allocation can
+        /// succeed even if there is not enough memory but then the test may get killed by the OOM killer at the
+        /// time the memory is accessed which triggers the full memory allocation.
+        /// Also see <see cref="Utf8JsonWriterTests.WriteLargeJsonToStreamWithoutFlushing"/>
+        /// </summary>
+        [PlatformSpecific(TestPlatforms.Windows | TestPlatforms.OSX)]
+        [ConditionalFact(nameof(Utf8JsonWriterTests.IsX64))]
+        [OuterLoop]
+        public static void SerializeLargeListOfObjects()
+        {
+            Dto dto = new()
+            {
+                Prop1 = int.MaxValue,
+                Prop2 = int.MinValue,
+                Prop3 = "AC",
+                Prop4 = 500,
+                Prop5 = int.MaxValue / 2,
+                Prop6 = 250.5M / 3,
+                Prop7 = 250.5M / 3,
+                Prop8 = 250.5M / 3,
+                Prop9 = 250.5M / 3,
+                Prop10 = 250.5M / 3,
+                Prop11 = 150.6M / 3,
+                Prop12 = 150.6M / 3,
+                Prop13 = DateTimeOffset.Now,
+                Prop14 = DateTimeOffset.Now,
+                Prop15 = DateTimeOffset.Now,
+                Prop16 = DateTimeOffset.Now,
+                Prop17 = 3,
+                Prop18 = DateTime.Now,
+                Prop19 = DateTime.Now,
+                Prop20 = 25000,
+                Prop21 = DateTime.Now
+            };
+
+            List<Dto> items = Enumerable.Repeat(dto, 3_000_000).ToList();
+            Assert.Throws<OutOfMemoryException>(() => JsonSerializer.SerializeToUtf8Bytes(items));
```

Comment:
nit: add a comment suggesting on approx how large the JSON would be just to make it clear we're not unnecessarily over-allocating on the 3,000,000 count for example.